base/reduce.jl

```diff
@@ -410,7 +415,13 @@ The default is `reduce_first(op, f(x))`.
 """
 mapreduce_first(f, op, x) = reduce_first(op, f(x))
 
-_mapreduce(f, op, A::AbstractArrayOrBroadcasted) = _mapreduce(f, op, IndexStyle(A), A)
+function _mapreduce(f, op, A::AbstractArrayOrBroadcasted)
+    fT = _return_type(f, Tuple{eltype(A)}) # determine type of f(A[i])
+    preproc = _makefast_preproc(op, fT) # fetch an accelerator, if available
+    ffast = preproc == identity ? f : preproc ∘ f # apply accelerator, if available
```

Comment:
```suggestion
    preproc = _makefast_preproc(op, fT) # fetch an accelerator, if available
    ffast = preproc === identity ? f : preproc ∘ f # apply accelerator, if available
```
Also, is this actually faster than
```suggestion
    _makefast_preproc(op, fT) ∘ f # apply accelerator, if available
```
<details>
<summary> This is what makes me question that </summary>

```julia
julia> f(x) = x^2
f (generic function with 1 method)

julia> using BenchmarkTools

julia> @belapsed map(identity ∘ f, 1:100)
7.436847710330139e-8

julia> @belapsed map(f, 1:100)
7.386489361702129e-8

julia> @belapsed map(identity ∘ f, 1:100)
7.409924487594391e-8

julia> @belapsed map(f, 1:100)
7.311134676564157e-8

julia> @belapsed map(identity ∘ f, 1:100)
7.403821656050955e-8

julia> @belapsed map(f, 1:100)
7.459745762711865e-8

julia> 

```

</details>

Reply:
There is no speed difference. However, the composition `identity \circ f` was breaking all the `reduce_empty` definitions. The ternary here keeps those from breaking in many cases, but I do worry it will still fail in some.

Reply:
I got some wires crossed. I'm implementing the second suggested change as well.